keps/sig-node/4191-split-image-filesystem/README.md

```diff
@@ -0,0 +1,891 @@
+
+### Stats Summary
+
+Stats Summary has a field called runtime and we will add a containerFs to the runtime field.
+
+```golang
+// NodeStats holds node-level unprocessed sample stats.
+type NodeStats struct {
+  ...
+// Reference to the measured Node.
+// Stats about the underlying container runtime.
+// +optional
+Runtime *RuntimeStats `json:"runtime,omitempty"`
+}
+
+// RuntimeStats are stats pertaining to the underlying container runtime.
+type RuntimeStats struct {
+// Stats about the underlying filesystem where container images are stored.
+// This filesystem could be the same as the primary (root) filesystem.
+// Usage here refers to the total number of bytes occupied by images on the filesystem.
+// +optional
+ImageFs *FsStats `json:"imageFs,omitempty"`
++ // Stats about the underlying filesystem where container's writeable layer is stored.
++ // This filesystem could be the same as the primary (root) filesystem or the ImageFS.
++ // Usage here refers to the total number of bytes occupied by the writeable layer on the filesystem.
++ // +optional
++ ContainerFs *FsStats `json:"containerFs,omitempty"`
+}
+```
+
+In this KEP, ContainerFs can either be the same as ImageFs or NodeFs.
+
+We will add a more detailed function for ImageFsStats in the Provider Interface
+
+```golang
+type containerStatsProvider interface {
+...
+ImageFsStats(ctx context.Context) (*statsapi.FsStats, *statsapi.FsStats, error)
+}
+```
+
+If we have a single image filesystem then ImageFs includes both writable and read-only layer.  In this case, `ImageFsStats` will return an identical object for ImageFs and ContainerFs.
+
+In a case where the container runtime does not return a container filesystem, we will assume that the image_filesystem=container_filesystem.
+This allows us kubelet to support container runtimes that have yet implemented the CRI implementation in `ImageFsInfo`.
+
+### CAdvisor
+
+Since CRI-O is still using CAdvisor Stats provider, it is necessary to support CAdvisor with separate disks.
+
+CAdvisor has plugins for each container runtime under containers. [CRI-O](https://github.com/google/cadvisor/tree/master/container/crio)
+
+The plugin in CRI-O relies on the endpoints `info` and `container/{id}`.  Info is used to get information about the storage filesystem and
+container gets information about the mount points.  CRI-O will add a new field for the `storage_image` to tell when we are splitting the filesystem. 
+
+This is used to gather file stats.  
+
+CAdvisor labels CRI-O images as `CRI-O-images` and that is assumed to be the mountpoint of the container.  When splitting the filesystem this
+ends up pointing to the writeable layer of the container.
+
+We will propose a new label in CAdvisor: `CRI-O-containers` will point to the writeable layer.  
+
+We have created [CAdvisor PR](https://github.com/google/cadvisor/pull/3395) to suggest how CAdvisor's can be enhanced to support a container filesystem.
```

Comment:
this piece is kind of CRI-O specific, is there a way we can generalize?

Reply:
updated.  We decided to just summarize what changes will be needed in both stats provider implementations (CRI Stats Provider and CAdvisor Stats Provider)